Add Runtime Fields Contexts to Painless Execute API

jdconrad opened the pull request:
This change adds support for the 7 different runtime fields contexts to the Painless Execute API. Each context can accept the standard script input (source and params) along with a user-defined document and an index name to pull mappings from. The results depend on the output of the runtime field type.

Closes https://github.com/elastic/elasticsearch/issues/70467

jdconrad committed:
add rt fields to painless execute action

jdconrad committed:
add some tests

jdconrad committed:
finish single node tests

jdconrad committed:
add yaml tests

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

jdconrad committed:
Merge branch 'master' into rtexec

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java:
```diff
@@ -105,7 +114,15 @@ private PainlessExecuteAction() {
                 "filter", FilterScript.CONTEXT,
-                "score", ScoreScript.CONTEXT);
+                "score", ScoreScript.CONTEXT,
+                "boolean_script_field_script_field", BooleanFieldScript.CONTEXT,
```
@nik9000 This "bug" with doubling the "script_field" suffix is unfortunate. WDYT about doing a hard break and fixing it? The only place the name is used is for stored scripts (but only in the api, to know which context to compile against) and in the node settings for script compilation limits, and with the latter it's not applicable here since unlimited is used.

nik9000 replied:
We disallowed stored scripts for runtime fields when we first built them and never re-allowed them. So I think that's mostly safe. I'm quite ok renaming things. I hadn't realized these names got out to users. Also - thing were different when we picked the names. We thought there'd be more different kinds of runtime fields. We never renamed them later. Anyway, what do you think of `boolean_field_script`? They are the script that runtime fields use for boolean fields *and* they are the script that we'll use to calculate boolean fields at index time. Same script context.

javanna replied:
I agree that it is odd to make users type this, but I would take this a step forward. I am not entirely comfortable even asking users to type context: "boolean_script_field" when they want to test a script for a boolean runtime field. It seems like there is a gap between how a runtime field is simulated and when it gets used in practice. There may be other ways to achieve this, but would it be a possibility to expose some artificial "runtime_field" context instead, and have another way to specify the type, for instance as part of the context setup? Any other ideas on this?

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java:
```diff
@@ -521,6 +538,76 @@ static Response innerShardOperation(Request request, ScriptService scriptService
+                    BooleanFieldScript booleanFieldScript = leafFactory.newInstance(leafReaderContext);
+                    booleanFieldScript.setDocument(0);
+                    booleanFieldScript.execute();
+                    return new Response(Map.of("trues", booleanFieldScript.trues(), "falses", booleanFieldScript.falses()));
```
This seems odd but I realize it is a result of the existing output of BooleanFieldScript. @nik9000 What about using a list of values like the other runtime field types do?

nik9000 replied:
I talked to @jdconrad about this yesterday and I don't have a great answer here. I like that `BooleanFieldScript` can do what it wants, at least on the theoretical level. And I don't think it's worth modifying `BooleanFieldScript` just for this API. The ordering of the emitted booleans isn't really important in any other context.

But I think it'd be entirely reasonable for this API to make a list out of the counts just to make it consistent. That is also consistent with the way doc values and the `fields` API work for this.

I do think its nice if we give these scripts the freedom to change their minds about they accumulate values without it breaking folks. I can't think of how they might do that right now but that don't mean there aren't good ways. If you "work like doc values" in this API then we have a consistent way to send the results to the callers.

As it stands now the scripts preserve ordering of values for everything but booleans. But the rest of ES leaves us the option to change that ordering in the future. And I'd like to keep it.

javanna replied:
I just stumbled upon the same as part of adding support for script to BooleanFieldMapper and I went for doing what doc_values do, meaning returning false values first, then true values.Given that doc_values are resorted, and the fields API does not guarantee ordering, it looks like it is not needed to hold an array of values for this specific case. What we could do is expose a method that returns the values in whatever order we decide to return them, so that consumers don't have to do it themselves. And I would not expose in the response the trues vs falses concept, but rather expose an array of values, regardless of whether we have it internally or not.

javanna replied:
Like I said, I don't think we should be exposing trues and falses here, but an array of values like we do for other types. I am adding a method to achieve the same [here](https://github.com/elastic/elasticsearch/pull/71454/files#diff-0938096fde41d31361acc4b51c62a1012aa99e97732d8f7159a1ca8fd9664aebR49), can we do something similar?

rjernst reviewed (approved):
LGTM but @nik9000 may have more thoughts.

nik9000 commented on modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/70_execute_painless_scripts.yml:
```diff
@@ -68,3 +74,108 @@ setup:
+            document:
+              rank: 4
+            index: "my-index"
+  - match: { result: [ 20 ] }
```
I think it'd be nice to make sure that multi-valued results work with this API. Folks'll use them less than single-valued results so we might not catch future bugs that break them until they are out in the wild.

jdconrad replied:
@nik9000 @javanna Based on the boolean feedback, when the results come back should I be ordering them to match what doc values would do then as well for others?

nik9000 replied:
Yeah - I'd sort like doc_values does. It might not be right, but at least it lines up with what we do in other places.

javanna replied:
I would add a method that does it to BOoleanFieldScript, I am doing the same as part of another task that is not a PR yet. you got to it first :)

javanna commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java:
```diff
@@ -521,6 +538,76 @@ static Response innerShardOperation(Request request, ScriptService scriptService
+                            factory.newFactory("boolean_runtime_field", request.getScript().getParams(), context.lookup());
+                    BooleanFieldScript booleanFieldScript = leafFactory.newInstance(leafReaderContext);
+                    booleanFieldScript.setDocument(0);
+                    booleanFieldScript.execute();
```
nit: we could call script.runForDoc(0)?

javanna replied:
can you address this please? does it make sense?

javanna reviewed (changes requested):
I left a couple of comments, the main one around how users will refer to the runtime field contexts.

jdconrad commented:
@javanna I agree about not wanting the user to type out that full name -- we could add an alias as you suggest for the names of the runtime field contexts. Maybe something like
```
"context": "runtime_field",
"context_setup": {
    "runtime_type": "long",
    ...
}
```

javanna commented:
++ @jdconrad I would only rename runtime_type to type in the context options ;)

jdconrad commented:
Just a note that @alisonelizabeth has https://github.com/elastic/kibana/pull/96424 ready to go for Painless Lab dependent on this PR.

rjernst commented:
>  I am not entirely comfortable even asking users to type context: "boolean_script_field

Users won't need to type this: that is entirely up to how Kibana exposes this. I do not think we should complicate the execute api (or any other scripting api) to make runtime fields special here. From the scripting level, runtime fields are many contexts. We need to expose those contexts. In fact, we still want to move the execute API out so any context can be included, programmatically, and doing so means associating the handling code for parsing the context setup to the named context. Having a non real "runtime_field" here would make that effort much more difficult, and unable to use the most natural api, keyed by the context.

So I think we should fix the names, but not try to coerce them any farther than necessary. Note that the other contexts do not have the "field" suffix. For example, score scripts are just "score". But I'm not opposed to keep field in this case, just thought I would point out this isn't something user friendly exactly to begin with, it is an opaque id.

jdconrad commented:
@javanna @rjernst @nik9000 I'd like to reach a consensus on this quickly if possible. The simplest change to make is what @rjernst has suggested. How would you all feel about renaming each of the runtime fields contexts to <type>_field ie `long_field` or `geo_point_field`, etc?

jdconrad commented:
@rjernst While we have used these as ids - I do wonder if it would make sense to also have aliases even given the added layer of complexity as part of the execute api. Although, I do agree that most users probably just go through Kibana and will never see this, it is still available as an external api.

javanna commented:
While the UI can help abstracting internal concepts away, I would prefer that we try to also simplify what concepts we expose to the direct users of our API. In this case, users know about runtime fields, and they specify a script and a type. They have no idea what a script context is and the fact that each type has its own. I don't think we should expose this internal detail in our API. I'd like to try and expose this one runtime field context for which you can specify the type as an additional configuration parameter. That is not perfect but close enough to how a runtime field is specified in the mapping or in the search request. @jimczi what do you think?

rjernst commented:
>  I don't think we should expose this internal detail in our API. I'd like to try and expose this one runtime field context for which you can specify the type as an additional configuration parameter. 

This API is not for runtime fields. It is an existing API, that has already defined semantics. That is, the context key matches a script context. Users will not be calling this API directly, they will be using a UI which will make a request to it behind the scenes, hidden to them. @javanna Can you clarify why given those facts how the context is specified *by Kibana* would matter?

alisonelizabeth commented:
To add some thoughts here in terms of the UI -

I imagine this API will be used in two places in Kibana:
1. Painless Lab (WIP PR - https://github.com/elastic/kibana/pull/96424) where we already have a notion of a user selecting an execution context. We can expose a user-friendly string for the context name (e.g., `boolean` instead of `boolean_script_field_script`).  

// Note the TODOs are just placeholder text for now.
<img width="1366" alt="Screen Shot 2021-04-07 at 12 33 24 PM" src="https://user-images.githubusercontent.com/5226211/113902040-7a277d00-979d-11eb-8670-512b2d0eea4d.png">

2. Runtime fields editor. I haven't been involved in recent discussions around this, but I don't foresee that we would need to expose the context to the user. We would determine what context to pass to the execute API based on the "type" the user selected.

jdconrad commented:
@elasticmachine run elasticsearch-ci/2

javanna commented:
I appreciate that we can hide internal details through the UI, but I don't think that it is a good reason not to make our API easier to use when we can. Painless execute is a public API, it is documented and can be used directly, hence I think it is more user-friendly to expose one runtime field context instead of seven. That simplifies the docs and the UI as well (otherwise it has to maintain the mapping between type and required context, and so does every other consumer of the API).

I cannot comment on the effort required, but my expectation was that this would require a small conversion layer, which should not be on the way of opening the API up to any context later on.

rjernst commented:
> which should not be on the way of opening the API up to any context later on.

The API is currently a hardcoded list of handlers, through a chain of if/else if looking at the context value specified. This hardcoded list is what we want to eventually change, moving the API outside of painless, so that any context created for scripting automatically gets supported. The way that would work is through the registration of scripting contexts that we already have (either directly in the context definition, or indirectly in the ScriptingPlugin interface). Using something *other* than the context id here will make such a transition difficult: it would mean defining more indirection layers (some "virtual" context called runtime_field and how that maps to concrete contexts).

> Painless execute is a public API,  it is documented and can be used directly, hence I think it is more user-friendly to expose one runtime field context instead of seven

Yes it is, but this API is specifically for scripting. Runtime fields are one use of scripting. I am strongly against trying to hide details of how the scripting api works. Context, while only exposed in a few places, is already exposed to users. We do lack good documentation on which contexts are available, but this is the means by which we have to define the signature of a type of script. Runtime fields has several because there are different signatures, depending on the type of the field.

> my expectation was that this would require a small conversion layer

The amount of work involved is not my concern, per se. It is the overhead of this conversion layer for both developers and users.  I find being direct is much simpler than adding indirection, especially when the *vast majority* of users will never see this, yet it would considerably complicate documentation (the API currently takes context, but now what is "context" if not the script context name?), and implementation (the indirection i already mentioned above).

jimczi commented:
> How would you all feel about renaming each of the runtime fields contexts to type_field ie long_field or geo_point_field, etc?

+1, I think it's inlined with the intent. We have strongly typed contexts for fields that produce values through a Painless script.
We added the support for these contexts in the indexed fields so I liked the idea of having generic typed contexts for the mapping.
The `context_setup` will be useful to add more customizations (additional runtime mapping for fields that depend on other scripts), although I don't think it should be used it to magically reconstruct the context name. These contexts exist so we need to expose them clearly, a simple name would help for that matter.

jdconrad commented:
@javanna Are you all right with us proceeding with specifying the full context name for now given this is how the execute api already operates, and we can revisit api specification for this at some point in the future if we get feedback that it's not working well?

javanna commented:
@jdconrad I am ok to not have a unified context for runtime fields and expose one per type. We were chatting with @jimczi and it turns out that the same contexts are also used for indexed fields that hold a script (see #68984), so using e.g. boolean_field would be good enough. Is then the plan to rename the existing contexts as a followup?

jdconrad commented:
@javanna Cool, thanks Luca! 

I will address the rest of Nik's and your requests to make sure the return values are in doc-values order.

And I do intend to follow this up with the name change, but after speaking with @nik9000 and @jimczi, it's possible users have added stored scripts with runtime field contexts, even though they're unusable. I would like to switch the names while also adding code that prevents this from happening and drops any that have been stored already with some kind of warning in the logs. (Please do let me know if you see an obvious flaw in this plan.)

rjernst commented:
> it's possible users have added stored scripts with runtime field contexts, even though they're unusable

Note that stored scripts don't store the context, it is only passed in to trigger compilation. So even if a stored script was passed in with the old name, changing the name does not break the script. It only breaks someone trying to use that name for stored scripts in the future.

jdconrad committed:
Merge branch 'master' into rtexec

jdconrad committed:
update based on pr feedback

jdconrad committed:
Merge branch 'master' into rtexec

jdconrad committed:
some yaml tests updated

jdconrad committed:
Merge branch 'master' into rtexec

jdconrad committed:
finish updating yaml tests

jdconrad commented:
@nik9000 Would you please take one last pass at this?

This last set of commits has the following:
* Updated context names to *_type (long_field, etc.)
* Updated to use the new asDocValues methods on the contexts for results
* Updated to use runAsDoc instead of execute
* Updated unit tests
* Updated yaml tests w/ single value and multi value for each type

nik9000 reviewed (approved):
LGTM. @javanna should have another look too.

jdconrad commented:
Thanks @nik9000. @javanna Would you please take one last pass at this? :)

jdconrad commented:
@elasticmachine run elasticsearch-ci/2

javanna reviewed (approved):
LGTM thanks @jdconrad !

jdconrad commented:
Thank you for the all the feedback, @javanna!
